opm/simulators/linalg/bda/BdaBridge.cpp

```diff
@@ -209,24 +255,74 @@ void BdaBridge<BridgeMatrix, BridgeVector, block_size>::solve_system(BridgeMatri
         res.converged = result.converged;
         res.conv_rate = result.conv_rate;
         res.elapsed = result.elapsed;
-    }else{
+    } else if (use_fpga) {
+        BdaResult result;
+        result.converged = false;
+        int *h_cols = nullptr, *h_rows = nullptr;
+        const int dim = (*mat)[0][0].N();
+        const int Nb = mat->N();
+        const int N = Nb * dim;
+        int nnz = mat->nonzeroes() * dim * dim;
+
+        if (dim != 3) {
+            OpmLog::warning("fpgaSolver only accepts blocksize = 3 at this time, will use Dune for the remainder of the program");
+            use_fpga = false;
+            return;
+        }
+
+        double *nnzsp = nullptr;
+        if (h_cols == nullptr) {
```

Comment:
Code duplication reduced in c65b2f5.
The `h_rows` and `h_cols` vectors store the sparsity pattern in contiguous memory. This can be removed if there is a `Dune::BCRSMatrix` that guarantees the same thing.
For every linear solve, it is checked whether the sparsity pattern is stored already.

Reply:
Great. Thanks.